src/test/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStepTest.java

```diff
@@ -176,6 +176,18 @@ public void scmVars() throws Exception {
         });
     }
 
+    @Test public void scmParsesChangelogFileFromFakeChangeLogSCM() {
+        rr.then(r -> {
+            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition(
+                    "import org.jvnet.hudson.test.FakeChangeLogSCM\n" +
+                            "node() {\n" +
+                            "  checkout(new FakeChangeLogSCM())\n" +
+                            "}", false));
+            r.buildAndAssertSuccess(p);
```

Comment:
Does the new test fail without the changes in `src/main`? I think it would still pass, right? To make it a regression test, you could add a line like [this](https://github.com/jenkinsci/workflow-job-plugin/blob/f61e0e7ee8bea00ac722f0fdb54f17bff150ebca/src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunTest.java#L381) to the Pipeline, and then add an assertion like this to the test:

```
WorkflowRun b = r.buildAndAssertSuccess(p);
assertThat(b.getCulpritIds(), Matchers.contains('alice1'));
```

Reply:
It did pass either way.

I chose to use [this Matcher](https://github.com/jenkinsci/workflow-scm-step-plugin/pull/40/commits/a575e8380bd262f5f04461a511c952178ddd5ce5#diff-862fb161c96b2514ed6613245a629e71R191) to look for `alice1`. [getCulpritIDs](https://github.com/jenkinsci/workflow-job-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L821-L834) returns `Set<String> [culprits]`(https://github.com/jenkinsci/workflow-job-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L183). Seems like there were a bunch of different ways to simply look for `alice1` so I picked this approach somewhat arbitrarily. Happy to do differently if there's a better way though.

Reply:
Your assertion is fine. Here's a brain dump on how I try to choose what kind of assertion is best in case it's helpful in the future:

In general, I try to keep the "actual" value passed to `assertThat` as rich as possible (e.g. don't flatten it to a `String` or `boolean`) and use matchers that give you the strictest guarantees that are correct for your case. For example, the simplest assertion would be something like this:

```
assertTrue(b.getCulprits().contains("alice1"));
// No special message if it fails, just a stack trace
```

Using `assertThat` and `containsString` like you did is better:

```
assertThat(b.getCulprits().toString(), containsString("alice1"));
// Failure message: `Expected: a string containing "alice1" but: was "[bob]"`
```

So the second assertion is better in that we can more easily see what went wrong when it fails, but both of the assertions are weaker than they could be. For example, neither would fail if something changed and there were two culprits instead of one, even though we probably want the test to fail in that case. We can avoid that issue by keeping the "actual" value a `Set` and then using the `equalTo` matcher with a `Set` that we supply: 

```
assertThat(b.getCulprits(), equalTo(Collections.singleton("alice1")));
// Failure message with no items:`Expected: <[alice1]> but: was <[]>`
// Failure message with wrong item:`Expected: <[alice1]> but: was <[bob]>`
// Failure message with more items: `Expected: <[alice1]> but: was <[alice1, bob]>`
```

We could also handle it using a fancier `Matcher`. Here's an example with a matcher that checks that an `Iterable` (the `Set` in our case) has all of the expected items in the given order (a stronger guarantee than `equalTo` in this case, which does not care about ordering) and does not have any additional items.

```
assertThat(b.getCulprits(), contains("alice1"));
// Failure message with no items: `Expected: iterable containing ["alice1"] but: No item matched: "alice1"`
// Failure message with wrong item: `Expected: iterable containing ["alice1"] but: item 0: was "bob"`
// Failure message with more items: `Expected: iterable containing ["alice1"] but: Not matched: "bob"`
```

Personally, I think that we don't really care about ordering in our test, and `equalTo` is easiest to understand at a glance and has the best error messages, so that's what I'd go with here.

Reply:
First of all that ^ is awesome.

Second - I noticed that you're using `getCulprits()` in those ^ examples, whereas I used `getCulpritIds()` in my assertion, based on your earlier suggestion. That made me curious and started experimenting with them both, using the `equalTo(Collections.siungleton("alice1"))` matcher. `getCulprits` produces a strange result. Says it doesn't match, but it clearly does.
```
➜  workflow-scm-step-plugin git:(check-timestamp-instead) ✗ mvn test -Dtest=SCMStepTest#scmParsesChangelogFileFromFakeChangeLogSCM
...
[ERROR] Failures: 
[ERROR]   SCMStepTest.lambda$scmParsesChangelogFileFromFakeChangeLogSCM$5:194 
Expected: <[alice1]>
     but: was <[alice1]>
```
Weird. Fails in my IDE too, even though the comparison viewer itself says "No differences." 
<img width="917" alt="image" src="https://user-images.githubusercontent.com/21689198/78204699-44540800-7468-11ea-8195-0635968a1bf6.png">. 

Both fail in the correct way when I switch [L126 of SCMStep](https://github.com/kshultzCB/workflow-scm-step-plugin/blob/ed9c65ad425887e53434cc7a69719a1ee5ff597a/src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java#L126) back to use [the old condition](https://github.com/jenkinsci/workflow-scm-step-plugin/blob/d5c06f9ae5a587ec46a6623fe91bab02514c4946/src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java#L125). Not sure what to make of that, just looks really strange to me.

In any case, I like what you said about the `equalTo` being more precise. I've stuck with `getCulpritIds()`, and switched to. It's all passed the PCT scenario, the tests in workflow-job also pass, and I think this should be the last one. Sorry to change this one last time, but it would bother me if I didn't.

I've learned a lot via this PR and I really appreciate that.

Reply:
> I noticed that you're using getCulprits() in those ^ examples, whereas I used getCulpritIds() in my assertion

Just a typo on my side, sorry!

Reply:
> getCulprits produces a strange result. Says it doesn't match, but it clearly does.

The problem is that `getCulprits()` returns `Set<User>`, not `Set<String>`. The failure message from `equalTo` goes through everything in the `Set` and calls `toString()` on it to display it, which is why the message is confusing, because [`User.toString()`](https://github.com/jenkinsci/jenkins/blob/8c1e567e7d400e2e51ab8f9bc7cf90ca8b0290c8/core/src/main/java/hudson/model/User.java#L690-L692) returns the name of the user, so the output looks like the same, but the types are different. Ideally the failure message would give some indication of the types on each side when they do not match to help clarify the issue, but as far as I know Hamcrest's matchers do not provide anything like that.